app/assets/src/components/views/amr_heatmap/AMRHeatmapVis.jsx

```diff
@@ -0,0 +1,262 @@
+    });
+  }
+
+  extractData(rawData) {
+    let genes = {};
+    let sampleLabels = [];
+    rawData.filter(rawSample => rawSample.error === "").forEach(sample => {
+      let sampleName = sample.sample_name;
+      sampleLabels.push({ label: `${sample.sample_name}` });
+      sample.amr_counts.forEach(amrCount => {
+        let gene = amrCount.gene;
+        let allele = amrCount.allele;
+        let depth = amrCount.depth;
+        let coverage = amrCount.coverage;
+        let genesDataStructure = {
+          name: gene,
+          samples: {
+            [sampleName]: {
+              name: sampleName,
+              depth: depth,
+              coverage: coverage,
+            },
+          },
+          alleles: {
+            [allele]: {
+              samples: {
+                [sampleName]: {
+                  depth: depth,
+                  coverage: coverage,
+                },
+              },
+            },
+          },
+        };
+        if (genes.hasOwnProperty(gene)) {
+          if (genes[gene].samples.hasOwnProperty(sampleName)) {
+            // do nothing for now
+          } else {
+            genes[gene].samples[sampleName] =
+              genesDataStructure.samples[sampleName];
+          }
+          if (genes[gene].alleles.hasOwnProperty(allele)) {
+            genes[gene].alleles[allele].samples[sampleName] =
+              genesDataStructure.alleles[allele].samples[sampleName];
+          } else {
+            genes[gene].alleles[allele] = genesDataStructure.alleles[allele];
+          }
+        } else {
+          genes[gene] = genesDataStructure;
+        }
+      });
+    });
+    return [genes, sampleLabels];
+  }
+
+  mapAllelesToGenes(geneData) {
+    let alleleToGeneMap = {};
+    Object.keys(geneData).forEach(gene => {
+      let alleles = geneData[gene].alleles;
+      Object.keys(alleles).forEach(allele => {
+        alleleToGeneMap[allele] = gene;
+      });
+    });
+    return alleleToGeneMap;
+  }
+
+  //*** Following functions must be called after the component has updated ***
+  //*** (i.e. after the component has requested AMR data and updated state) ***
+
+  createGeneLabels() {
+    let geneData = this.state.geneData;
+    let geneLabels = [];
+    Object.keys(geneData).forEach(gene => {
+      geneLabels.push({ label: `${gene}` });
+    });
+    return geneLabels;
+  }
+
+  createAlleleLabels() {
+    let geneData = this.state.geneData;
+    let alleleLabels = [];
+    let genes = Object.keys(geneData);
+    genes.forEach(gene => {
+      let geneAlleles = Object.keys(geneData[gene].alleles);
+      let labeledAlleles = geneAlleles.map(allele => {
+        return { label: `${allele}` };
+      });
+      alleleLabels = alleleLabels.concat(labeledAlleles);
+    });
+    return alleleLabels;
+  }
+
+  createHeatmapLabels() {
+    let viewLevel = this.state.viewLevel;
+    switch (viewLevel) {
+      case METRIC_ALLELES: {
+        return this.createAlleleLabels();
+        break;
```

Comment:
nit: the return statement should exit out of the function already, so no need for break